Sources on Multiple Workers - Intermediate Version

JONBRWN opened the pull request:
**DO NOT MERGE**

This PR is open only for review and is not intended to be merged to `master` in its current state.

## Testing

Testing to confirm that sources start on multiple workers can be done using the `multi_partition_detector` application.

**Manual Testing**

1. `cd testing/correctness/apps/multi_partition_detector`
2. `make build build-giles-sender build-utils-data_receiver`
3.Start Data Receiver: ```./utils/data_receiver/data_receiver -l 127.0.0.1:45674 -f > received.txt```
4. Start Initializer: ```./multi_partition_detector --in Detector@127.0.0.1:20000 --out 127.0.0.1:45674 --metrics 127.0.0.1:40614 --name initializer --worker-count 2 --data 127.0.0.1:20002 --external 127.0.0.1:20013 --cluster-initializer  --control 127.0.0.1:20015   --ponythreads=1 --ponypinasio --ponynoblock```
5. Start Worker 2: ```./multi_partition_detector --in Detector@127.0.0.1:20010 --out 127.0.0.1:40757 --metrics 127.0.0.1:46008 --name worker1   --my-control 127.0.0.1:20004 --my-data 127.0.0.1:20005 --external 127.0.0.1:20006 --control 127.0.0.1:20015   --ponythreads=1 --ponypinasio --ponynoblock```
6. Start Sender for `key_0`: ```giles/sender/sender -h 127.0.0.1:20000 -s 1 -i 50_000_000  --ponythreads=1 -y -g 12 -w -u -m 100 -p key_0```
7. Start Sender for `key_1`: ```giles/sender/sender -h 127.0.0.1:20010 -s 1 -i 50_000_000  --ponythreads=1 -y -g 12 -w -u -m 100 -p key_1```
8. Run `validator`: ```./testing/correctness/apps/multi_partition_detector/validator/validator -i received.txt ```

**Automated Testing**
1. `cd testing/correctness/apps/multi_partition_detector`
2.Run tests ` make integration-tests-testing-correctness-apps-multi_partition_detector`
3. Run tests with resilience ` make integration-tests-testing-correctness-apps-multi_partition_detector`

**NOTE:** Currently the 10 worker tests fail when resilience is on with the following error:

```
ExternalChannelConnectNotifier (initializer): server closed
ExternalChannelConnectNotifier (initializer): server closed
Detector source: accepted a connection
worker_ack_barrier from InProgressPrimaryBarrierHandler
worker_ack_barrier from InProgressPrimaryBarrierHandler
worker_ack_barrier from InProgressPrimaryBarrierHandler
TCPSource connection closed
Detector source: accepted a connection
Received CheckpointBarrierToken(2) when still processing CheckpointBarrierToken(3) at step 172758834876403722331532747725528212546
This should never happen: failure in /home/vagrant/dev/wallaroo/lib/wallaroo/ent/barrier/barrier_step_forwarder.pony at line 79
```

jtfmumm commented:
Can we point this PR at a branch that will be used for the 2tier->1tier project?

JONBRWN commented:
@jtfmumm do you know which branch that is?

jtfmumm commented:
@JONBRWN Create a new branch of `master` and point this PR at that one.  It's a branch that represents the intermediate version of this feature.  Simon can then use that branch for the 2tier->1tier work.

jtfmumm commented:
I think it might make sense to add a type alias to source.pony:
```
type SourceName is String
```
and use that everywhere you're referring to the source name in the code (following the pattern of `WorkerName`).  That will make it easier to see at a glance what's happening.

JONBRWN commented:
@jtfmumm pointing to a new branch, will update `SourceName` to a type

JONBRWN commented:
@jtfmumm I've added the `SourceName` type and updated the relevant files. let me know if you think it'd make sense to update all references in Wallaroo of `source_name: String` etc to this as part of this PR or not

jtfmumm commented:
@JONBRWN Yes, I think it makes sense to update all instances that are related to the new source name concept, including the Map[String, Whatever] examples.

JONBRWN commented:
@jtfmumm updated the maps as well

jtfmumm commented:
@JONBRWN Ok, I've left more comments and one or two haven't been addressed.

jtfmumm commented:
You can cherry-pick `4f745f1e85333c7681794020ddf81d7f26294e57` to get the fix for that failing integration test.

JONBRWN commented on lib/wallaroo/core/source/gen_source/gen_source_config.pony:
```diff
@@ -23,12 +23,23 @@ use "wallaroo/core/source"
+  let _worker_source_config: WorkerGenSourceConfig[In]
 
   new val create(gen: GenSourceGeneratorBuilder[In]) =>
     _gen = gen
```
We do, for the call on line 36

jtfmumm replied:
38 doesn't use `_gen`, at least not in the code shown here

jtfmumm replied:
line 30 can use `gen`

JONBRWN replied:
@jtfmumm ignore my original comment, I thought you were referring to line 26 not line 29

JONBRWN commented on lib/wallaroo/core/source/source.pony:
```diff
@@ -29,10 +29,10 @@ use "wallaroo/core/metrics"
 interface val SourceConfig
   fun default_partitioner_builder(): PartitionerBuilder
-  fun source_listener_builder_builder(): SourceListenerBuilderBuilder
+  fun val source_listener_builder_builder(): SourceListenerBuilderBuilder
```
it's needed for the function in `tcp_Spource_config.pony:97` due to the use of `this`

jtfmumm replied:
ok

JONBRWN commented on lib/wallaroo/wallaroo_config.pony:
```diff
@@ -16,13 +16,16 @@ Copyright 2017 The Wallaroo Authors.
 class StartupOptions
   var m_arg: (Array[String] | None) = None
-  var input_addrs: Array[Array[String]] val = recover Array[Array[String]] end
+  var input_addrs: Map[String, (String, String)] val =
```
@jtfmumm let's talk about this one, as it existed here previously but from my understanding it's not used in the code

JONBRWN commented:
Confirmed that both GenSource and TCPSource 10 worker tests are now passing, squashing and merging
